c++/src/capnp/compiler/capnp.c++

```diff
@@ -50,7 +50,9 @@
 
 #if _WIN32
 #include <process.h>
+#ifndef WIN32_LEAN_AND_MEAN
```

Comment:
I understand `WINVER` and `_WIN32_WINNT`, but do we expect a build system might define `WIN32_LEAN_AND_MEAN`, too? This seems wrong, since this macro actually hides APIs. What if some module in KJ needs to use an API that is not part of the "lean and mean" set?

Reply:
For better or worse it's defined globally for all Windows builds @ Oculus. If some module in KJ actually needed that API there are 3 options:
1. Put code into kj to `#ifndef/#undef` up-front
2. Make it the responsibility of the external build system to exclude it for that 1 file

2 would be my preference and generally compilers will always take the last definition so that seems like a tractable solution assuming it's a built-time issue. If you wanted to be extra clear with the compiler error message you could `#ifndef WIN32_LEAN_AND_MEAN`/`#error`.

That being said, I *think* the conflict really comes from the few spots where `WIN32_LEAN_AND_MEAN` is defined as `1` instead of just defined at all. It would solve my problem if we removed those to avoid the preprocessor conflict. That being said, it wouldn't be unreasonable for a build maintainer in the general case to have defined `WIN32_LEAN_AND_MEAN` to 1 (clearly even within kj it was done both ways).

I don't have any strong opinions here. Let me know your preference.

Reply:
I would argue that it's incorrect to define this globally when building other people's code.

That said, if all of KJ builds correctly with this defined, then great. How about we move the whole ifdef/define into `win32-api-version.h`, and make sure to include that in all the places that currently define WIN32_LEAN_AND_MEAN? That would reduce the amount of boilerplate overall.

Reply:
Will do. I don't disagree with your point. I think it will be challenging though to figure out the best way to enable that policy as I'm sure there are good reasons it was done. The build team here is super sharp. The build macros do know if the code is third party or not so maybe keying off of that to add `/UWIN32_LEAN_AND_MEAN` might be a generic way to do that. On the other hand I suspect it could be a policy thing to reduce build capacity given how limited our Windows build farm is.

I'll apply your recommendations to this diff hopefully today.